arrows/vtk/applets/color_mesh.cxx

```diff
@@ -0,0 +1,527 @@
+      "We use threshold >= 0 to fix floating point inaccuracies "
+      "Default value is 0, bigger values will remove more points.");
+    config->set_value(
+      "remove_occluded", true,
+      "Remove occluded points if parameter is true.");
+    config->set_value(
+      "remove_masked", true,
+      "Remove masked points if parameter is true.");
+
+    kva::video_input::get_nested_algo_configuration("video_reader", config,
+      kva::video_input_sptr());
+    kva::video_input::get_nested_algo_configuration("mask_reader", config,
+      kva::video_input_sptr());
+    return config;
+  }
+
+  void initialize()
+  {
+    kva::video_input::set_nested_algo_configuration(
+      "video_reader", config_, video_reader_);
+    kva::video_input::set_nested_algo_configuration(
+      "mask_reader", config_, mask_reader_);
+  }
+
+  static kv::camera_map_sptr load_camera_map(
+    kva::video_input_sptr video_reader, std::string const& video_file,
+    std::string const& cameras_dir)
+  {
+    if ( video_reader == nullptr )
+    {
+      return nullptr;
+    }
+
+    video_reader->open(video_file);
+
+    kv::metadata_map_sptr metadata_map = video_reader->metadata_map();
+    vital::camera_map::map_camera_t cameras;
+
+    for (auto const& frame_metadata : metadata_map->metadata() )
+    {
+      const kv::metadata_vector mdv = frame_metadata.second;
+      const size_t frame_ID = frame_metadata.first;
+      const std::string name = basename_from_metadata(mdv, frame_ID);
+      try
+      {
+        cameras[frame_ID] = vital::read_krtd_file( name, cameras_dir );
+      }
+      catch ( const vital::file_not_found_exception& )
+      {
+        continue;
+      }
+    }
+
+    if ( cameras.empty() )
+    {
+      VITAL_THROW( vital::invalid_data, "No krtd files found" );
+    }
+
+    video_reader->close();
+    return vital::camera_map_sptr( new vital::simple_camera_map( cameras ) );
+  }
+
+  static vtkSmartPointer<vtkPolyData> load_mesh(std::string const& input_mesh_file)
+  {
+    std::string ext = ST::GetFilenameExtension(input_mesh_file);
+    if (ext == ".ply")
+    {
+      vtkNew<vtkPLYReader> reader;
+      reader->SetFileName(input_mesh_file.c_str());
+      reader->Update();
+      vtkSmartPointer<vtkPolyData> mesh = reader->GetOutput();
+      return mesh;
+    }
+    else if (ext ==  ".obj")
+    {
+      vtkNew<vtkOBJReader> reader;
+      reader->SetFileName(input_mesh_file.c_str());
+      reader->Update();
+      vtkSmartPointer<vtkPolyData> mesh = reader->GetOutput();
+      return mesh;
+    }
+    else if (ext == ".vtp")
+    {
+      vtkNew<vtkXMLPolyDataReader> reader;
+      reader->SetFileName(input_mesh_file.c_str());
+      reader->Update();
+      vtkSmartPointer<vtkPolyData> mesh = reader->GetOutput();
+      return mesh;
+    }
+    else
+    {
+      return nullptr;
+    }
+  }
+
+  void save_mesh(vtkSmartPointer<vtkPolyData> mesh, const char* output_path)
```

Comment:
Did you mention that you are going to write `.obj` and `.ply` as well?

Reply:
I don't think so - this is what was originally supported by TeleSculptor. We can certainly add it though.

Reply:
@danlipsa please add support for writing to `.ply`. or `.vtp`.  You can decide which writer to use based on provided file extension.  `.ply` is a common output format for TeleSculptor because it can be opened by many more 3rd party programs than `.vtp`.  `.ply` supports color per vertex which we used here when we compute the median or mean color.  `'ply` will not support saving colors for multiple frames.  That was a feature requested by one customer and is not commonly used.  We can skip `.obj` for now because it does not support color per vertex.  However, if we ever get around to using texture maps instead of color per vertex, `.obj` supports texture maps, and I believe `.ply` does not.

Reply:
Sounds good. I'll add PLY writing only for the average color. If the user wants PLY with color for multiple frames I'll print an error saying to use VTP.

Reply:
Or maybe just save it as VTP and print a warning.

Reply:
Done